File: Algorithms/python/107-BinaryTreeLevelOrderTraversalII/BinaryTreeLevelOrderTraversalII.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        level_list = list()

        if root is None:
            return level_list

        node_q = list()
        node_q.append(root)
        node_size = 1

        while node_size > 0:

            current_list = list()

            for i in range(node_size):
                node = node_q.pop(0)
                current_list.append(node.val)

                if node.left is not None:
                    node_q.append(node.left)
                if node.right is not None:
                    node_q.append(node.right)
            level_list.append(current_list)
            node_size = len(node_q)

        level_list.reverse()
        return level_list

File: Algorithms/cpp/BinaryTreeLevelOrderTraversalII.cpp
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        vector <vector<int> > ans;
        if (root == NULL)
            return ans;

        queue<TreeNode*> levelQueue;
        levelQueue.push(root);

        while (!levelQueue.empty()) {
            int sizeOfQueue = levelQueue.size();
            vector<int> levelVals;

            for (int i = 0; i < sizeOfQueue; i++) {
                TreeNode* node = levelQueue.front();
                levelQueue.pop();
                levelVals.push_back(node->val);
                addBackIfNotNull(levelQueue, node->left);
                addBackIfNotNull(levelQueue, node->right);
            }
            ans.push_back(levelVals);
        }
        reverseVector(ans);
        return ans;
    }

private:
    void addBackIfNotNull(queue<TreeNode*>& q, TreeNode* tn) {
        if (tn) {
            q.push(tn);
        }
    }
    void reverseVector(vector<vector<int>>& v) {
        int len = v.size();
        for (int i = 0, j = len - 1; i < j; i++, j--) {
            vector <int> t = v[i];
            v[i] = v[j];
            v[j] = t;
        }
    }
};
